### ISSUE TITLE
Bounded lists and strings

### ISSUE THREAD

**Issue by cpetig** [WebAssembly/component-model#385]
This extends #181 with the ability to embed bounded strings or lists directly within other structures while maintaining the variable length property.

Some environments require avoiding all memory allocations at runtime. This happens typically in either small embedded environments or safety related products.

To achieve this in a call returning a string or list/vector you typically combine a fixed size buffer with a length indicator into a containing data structure, e.g. see [heapless::String](https://docs.rs/heapless/latest/heapless/struct.String.html) or [boost::static_string](https://www.boost.org/doc/libs/1_85_0/libs/static_string/doc/html/index.html).

Thus I propose to optionally extend the fixed size buffers of #181 as implemented in #384 with a preceding length indicator. This length indicator could be of type u8 for strings/vectors below 256 elements, u16 below 64k and u32 otherwise. Unused elements in the buffer should be considered uninitialized (`MaybeUninit` in Rust).

This addresses the problem described in my recent comment in #383 .

I propose the WIT syntax `string<..32>` or `list<u32, ..16>`, the binary format would need a new type for bounded strings, lists could use a negative len specifier in 0x67 (see #384 ) or a different defvaltype.

**Comment by lukewagner** [comment]
This is a good idea and I've been considering it too.  Assuming we add the readable-/writable-buffer types proposed at the end of #369, are there current use cases you're looking at that would benefit from this over the buffer types?  Most use cases I can think of could alternatively use readable-/writable-buffer types just as well, although I can imagine some more complex hypothetical scenarios that couldn't.

**Comment by cpetig** [comment]
Whether lazy value lowering can support returning into pre-allocated buffers equally well is a good question. 

I know that if caller and callee share the same maximum size the string/list can never overflow the provided buffer on the callee side, simplifying the code considerably.

But what really sold me to *this* solution is the ease of use on the caller API side. E.g. `f: func() -> string<..16>;` can be used as `let value = f();` (value being of type `heapless::String<16>` instead of separately providing a buffer and handling overflow. I simply have no idea how to call lazy value lifting in Rust or C++ with the same elegance.

**Comment by lukewagner** [comment]
Yes, good point; bounded lists/strings would have the ideal source bindings for languages with facilities for inline allocation.

In the motivation given in your first post, you mentioned the common pattern in C-style APIs of transferring an arbitrary-sized value by repeatedly calling a function and transferring it out chunk at a time.  But for that specific case, I would think the writable-buffer type would be the better choice in that it both: (1) gives the caller control of the buffer size (instead of fixing it in the WIT once and for all), (2) provides a pointer to the linear memory up-front as a parameter which, depending on the host implementation, may remove a copy.

However, that's not the only use case for bounded lists/strings, so I mostly wanted to ask, for prioritization reasons, if you had other specific use cases in mind (in WASI or other WIT interfaces you're looking at).

**Comment by cpetig** [comment]
My primary motivation for working on this is the (functional safety) requirement of deterministic execution time. The most straightforward solution is with dedicated pre-allocated buffers provided by the caller. 

[Return value optimization](https://en.wikipedia.org/wiki/Copy_elision#Return_value_optimization) can enable this even via the very friendly return notation (which internally maps to an output parameter for non atomic types).

I feel with #384 ready (and having its own merit) and this being just a small extension this would be the lowest hanging fruit, given that `stream` and `writable-buffer` feel much less possible to finalize within this year.

**Comment by lukewagner** [comment]
Right, doing this as a small addition to #384 makes sense and I think the feature makes sense in general.  In terms of scheduling/prioritization, you're right that `stream` will take a while, but I had been thinking that maybe it'd make sense to work on readable/writable buffers as the very next thing.  That being said, fixed+bounded-length lists are definitely the smaller/simpler feature and so we could have them ready somewhat sooner -- but I wonder whether the (embedded and other) use cases that want "caller-supplied buffers" will be satisfied or will simply immediately turn to wanting readable/writable buffers.

And just, concretely, if we did have bounded-length lists, but not yet buffers, to avoid `realloc` with `read`, we'd need something like:
```wit
  read: func(n: u64) -> result<list<u8; ..128>>;
```
and I wonder if that hard-coded constant (whatever we choose) will seem too ad hoc and imminently-meant-to-be-deprecated to actually ship as a whole WASI release, knowing that what `read` really wants is:
```wit
  read: func(buf: writable-buffer<u8>) -> result;
```
OTOH, depending on the timings involved, it could be practically worth it to ship the former before the latter.  ¯\_(ツ)_/¯

**Comment by cpetig** [comment]
To be honest I never had adding limits to WASI in mind when I thought of this feature, but more specialized custom WIT interfaces which are fully controlled by the runtime designer (or used between two components where the interface can be unilaterally defined by a single party).

I get that this intention defeats the purpose of a generic standardized interface, which is exactly what WASI is about. But for now bounded lists seem the best option for user friendly zero allocation interfaces.

**Comment by lukewagner** [comment]
I see, thanks for explaining.  Given that #181 was primarily waiting for implementation feedback before merging and this is a relatively small addition, and you've helpfully created #384 as a rebase of #181, would you like to also add this idea to #384?  Given both options, perhaps the WAT syntax we want is `(list u8 (eq 4))` and `(list u8 (lt 4))` for `list<u8; 4>` and `list<u8; ..4>` resp.

**Comment by cpetig** [comment]
Oh, wow, this sounded like a small thing to do, but I found that the following things make it really large:
 - `definitions.py` has no way to create dummy data or read over data, so `lift_flat` and `lower_flat` for bounded types ended up much more laborious
 - String encoding creates a challenge to flat argument passing, I settled for utf8 or utf16 values per argument, `latin1+utf16` makes the length indicator more complex (I multiplied by four to handle the double amount of entries for latin1 and have an extra bit for the tag)

Also I went for an additional byte in the 0x67 encoding to distinguish between fixed lists, bounded lists and bounded strings.

You can find the current prototype at https://github.com/cpetig/component-model/tree/bounded-lists .
I will continue the work, but it will take longer than originally expected.

**Comment by lukewagner** [comment]
Oh right, I forgot to get back to strings.  Because of the encoding issues, I think perhaps we should not attempt to extend strings.  Instead, if an API wants to do this bounded-size thing, it can just pick a particular encoding (as part of the API's contract) and use a `list<u8; ..N>`.  Otherwise, we'd need `string<..K>`, where `K` is necessarily the number of Unicode Scalar Values, to imply a `K*4`-byte buffer which would be a lot more wasteful than a UTF-8-specialized `list<u8; ..K>`, which simply implies a `K`-byte buffer.

**Comment by cpetig** [comment]
I don't think there is a perfect solution for strings due to encoding variation, but taking the maximum length as the number of u8s in utf8 and number of u16s in any of the two utf16 encodings, there is still some possibility of overflow when recoding between components - but the bound has an easy to understand meaning and if both components have the same encoding the bound is identical, so no overflow is possible in the more common case.

At least this was the most sensible solution I came up with.

**Comment by badeend** [comment]
In addition to an upper bound, could we also add a lower bound, like `list<u8, 1..255>`?

**Comment by lukewagner** [comment]
Sounds reasonable to me.

**Comment by cpetig** [comment]
I understand that from a contract design standpoint a lower bound makes sense, but I can't come up with a practical example where it would be beneficial. @badeend Do you have a specific case in mind?

**Comment by badeend** [comment]
Indeed to indicate a list may not be empty:

---

https://github.com/WebAssembly/wasi-io/blob/285a31442f687a5ceee86044f9f2b0a9e2beae21/wit/poll.wit#L34C4-L46C57
```wit
    /// This function traps if either:
    /// - the list is empty, or:
    /// - the list contains more elements than can be indexed with a `u32` value.
    ///
    /// A timeout can be implemented by adding a pollable from the
    /// wasi-clocks API to the list.
    ///
    /// This function does not return a `result`; polling in itself does not
    /// do any I/O so it doesn't fail. If any of the I/O sources identified by
    /// the pollables has an error, it is indicated by marking the source as
    /// being ready for I/O.
    @since(version = 0.2.0)
    poll: func(in: list<borrow<pollable>>) -> list<u32>;
```

Better expressed as:

```wit
    poll: func(in: list<borrow<pollable>, 1..>) -> list<u32>;
// or even:
    poll: func(in: list<borrow<pollable>, 1..4294967295>) -> list<u32>;
// :P
```

---

https://github.com/badeend/wasi-sockets/blob/8f3b082e6ead0eebbcedf10d8a06d16dffbaa68d/wit/tls.wit#L30-L35
```wit
    /// Application-Layer Protocol Negotiation (ALPN) protocol ID.
    ///
    /// ALPN IDs are between 1 and 255 bytes long. Typically, they represent the
    /// binary encoding of an ASCII string (e.g. `[0x68 0x32]` for `"h2"`),
    /// though this is not required.
    type alpn-id = list<u8>;
```

Better expressed as:

```wit
    type alpn-id = list<u8, 1..255>;
```

**Comment by flelchuk** [comment]
Totally appreciate that fixed size lists have been merged!
I believe these bounded lists `list<T, ..N>` could be another great addition towards avoiding some allocations and intent-explicit interface definitions.

Although it seems like just a minor addition of a `length` value at first glance, we cannot easily emulate this even with fixed size lists.
In some numeric applications, unpopulated elements could possibly be set `NaN` or zero in a fixed size list.
Generally, one would have to write something like `tuple<list<option<T>, N>, u32>` - a completely unwieldy type in WIT and source languages, lacking length consistency enforcement on binding level.
With bounded lists we could generate bindings with idiomatic Rust types like `&[T]` / `heapless::Vec` or `arrayvec::ArrayVec`.

I've read through some conversations about caller provided buffers and related concepts - and bounded size lists seem like a much lower bar. Maybe we can have the bounded list as regular container/data type for the time being and offer caller provided buffers as a more optimized/expert approach later?

I would volunteer to join implementation efforts if/once there is consensus and WIT syntax is accepted.

**Comment by cpetig** [comment]
> I would volunteer to join implementation efforts if/once there is consensus and WIT syntax is accepted.

I think the existing syntax "consensus" is:
- Dynamic: `list<T>`/`(list t)`
- Fixed Size: `list<T, N>`/`(list t n)`
- (new) Bounded, potentially empty: `list<T, ..N>` or `list<T, 0..N>` - or should we go for `list<T, 0, N>` instead? I guess `(list t (0 n))` is the only reasonable wat syntax.
- (new) Bounded, at least one element: `list<T, 1..N>` or `list<T, 1, N>`
- (new) Dynamic, with guaranteed minimum length: `list<T, 1..>` (I can't come up with a good syntax for the comma variant). `(list t (1 none))`?

I have the feeling that a higher minimum length than 1 is rarely needed, so this would add five new binary types (string_0_N, string_1_N, list_0_N, list_1_N, list_1).

For bounded strings, I haven't sensed any good consensus on the interpretation of the bound. I now think that a more meaningful limit is (always 2 byte aligned, even number of) bytes, not characters, although this means that values could become un-representable when re-encoded between components, but at least data structures don't change size depending on encoding.

**Comment by flelchuk** [comment]
To be honest, `list<T, ..N>` is my favorite notation and introducing a lower bound does feel weird:

1. at first glance `list<T, 1..N>`, one might think the range are indices (i.e. source language arrays would start at 1)
2. this would probably be the first (explicit) occurence of WIT bindings enforcing contracts (list shall have no *less* than x items at runtime).
    ... well maybe transcoding strings can be viewed as a contract check while doing the work
    ... and discriminant matching
   but those are quite implicit.

Maybe the lower bound is better introduced at a point when there is consenus to add more runtime checks throughout, so that bounded lists do not appear overengineered compared to other parts. Brings back memories of Pascal subrange types (`type TDiceRoll = 1 .. 6;`)

About bounded strings, I did not quite get the reason for 2-byte alignment? We could define `string<..N>` such that N is the number of guaranteed available codepoints? In particular, `string<N>` and `string<..N>` could be sugar for `list<char, N>` and `list<char, ..N>` where a WIT `char` is 32-bit fixed-width. Not compact but at least deterministic and intelligible. Also matches `[char; N]` in Rust.
Bindings would probably want to represent this not as a plain `[char; N]` but `arrayvec::ArrayString<4*N>` to get a much more versatile `&str` (USVs are 1..4 bytes).

**Comment by rossberg** [comment]
Drive-by comment: Bounds (both lower and upper) are primarily _semantic_ constraints. But IIUC, the purpose of the component type system is to describe representations, not semantics. A fixed length can easily be argued to be a representation aspect (e.g. distinguishing static vs dynamic data structures), but bounds much less so.

**Comment by cpetig** [comment]
> To be honest, `list<T, ..N>` is my favorite notation and introducing a lower bound does feel weird:

It would be possible to standardize `list<T, ..N>`  first and add the rest once we found a consensus on a syntax. I feel strings open an even larger can of worms and thus should be de-prioritized.

For a wat syntax `(list s32 (le 8))` looks like a reasonable suggestion.

> About bounded strings, I did not quite get the reason for 2-byte alignment?

That idea kept the encoding option, so utf16 strings would be two byte aligned and should have an equal number of bytes. I like the `list<char, ..32>` compromise, although it clearly wastes space.

**Comment by flelchuk** [comment]
Fully agree. If there is need, strings can be added another day. Once fixed and bounded length lists are both there one could still add strings as a specialization with ` T=char`  with lower innovation effort, or, there may be better ideas by that time.

**Comment by cpetig** [comment]
> Drive-by comment: Bounds (both lower and upper) are primarily _semantic_ constraints. But IIUC, the purpose of the component type system is to describe representations, not semantics. A fixed length can easily be argued to be a representation aspect (e.g. distinguishing static vs dynamic data structures), but bounds much less so.

Sorry, it took me some time to understand the question. Indeed I am aiming for a representation which works without allocation and can be stored in line (e.g. inside a struct or contiguous memory region). The storage is reserved and wastes a lot of space and isn't cache friendly, but this is common design for functional safety. This is mostly to map existing solutions to WIT without going full into flat representation in shared memory buffers (which is also bounded but cache friendly), see https://github.com/WebAssembly/component-model/issues/398 for the more complex solution.

**Comment by Felix-El** [comment]
Syntax `list<T, ..N>` is as good as anything and I would really wish for it to land. With fixed-size lists accepted this seems well in reach, too. This has been open more than 18 months without too much debate.

@cpetig maybe this could make further progress if you resurrected your prototype at https://github.com/cpetig/component-model/tree/bounded-lists (minus the bounded strings, maybe binary encoding adjusted) and created a PR

**Comment by Felix-El** [comment]
Just created the PR from a rebase of cpetig's original prototype with minor modifications and less the bounded strings.

**Comment by rossberg** [comment]
@cpetig, sorry for the late response. I'm afraid I still don't quite see the justification for adding bounded lists. I understand that they are for pre-allocation, but it seems to me that is conflating abstract value representations with lifting/lowering concerns. Concretely, such a list is not functionally different from a variable list, and will contain the same dynamic information. The type encodes some additional semantic meta-property about its expected length, but that bound could just as well be out-of-band knowledge.

The way I see this, it is just a special case of wanting a non-canonical lifting/lowering for variable-length lists. But there are plenty of use cases that would urgently want non-canonical liftings/lowerings, for all sorts of types, and clearly, they can't all be built in as separate "canonical" types. Making an exception and privilege for this specific case by elevating it feels like cutting corners. The correct solution would be the ability to define custom lowerings, which IIUC is a long-term goal anyway.
